### PR TITLE
IFTI with Alias Templates.

### PR DESCRIPTION
**Pull request by Lucipetus**
This PR tries to enhance IFTI with certain amount of capability to resolve the parameters of alias templates.

Previous attempts include [DIP1023](https://github.com/dlang/DIPs/blob/master/DIPs/other/DIP1023.md).

Some of the related issues (all fixable with this PR):
[16465](https://issues.dlang.org/show_bug.cgi?id=16465)
[16486](https://issues.dlang.org/show_bug.cgi?id=16486)

# The problem

The following code doesn't compile.

```D
struct Matrix(U, size_t M, size_t N) {}
alias Vector(U, size_t N) = Matrix!(U, N, 1);
void foo(U, size_t N)(Vector!(U, N) v) {}
void main()
{
 foo(Vector!(float, 3)()); // error
 
 import std.regex: Regex;
 import std.traits: isInstanceOf;

 static assert(isInstanceOf!(Vector, Vector!(float, 3))); // error
 static assert(isInstanceOf!(Regex, Regex!char)); // error.
}
```

The cause is, D only tries to match the template declarations (in this case, `Vector` to `Matrix` ), it of course fails and returns nomatch.

# The solution

We can introduces an extra procedure after failing:

1. find the alias declaration (`alias Vector(U, size_t N) = Matrix!(U, N, 1)`)) and the type `Matrix!(U, N, 1)`.
2. call `deduceType` with (`[U, N]`, `Matrix!(U, N, 1)`, and the target instance `Matrix!(float, 3, 1)`).
3. the output `[float, 3]` can then be matched to the `parameters` in this frame, with largly the same parameter matching algorithm existing.*\
<del>*In fact the current implementation is a copy and paste, with minor alterations and simplifications (just to prove the idea).</del>\
*Current implementation reuses the old matching code.

This works recursivly on a chain of aliases. e.g.

```D
struct Matrix(U, size_t M, size_t N)
{

}

alias Vector(U, size_t N) = Matrix!(U, N, 1); // 1. resolves [U, N] > [float, 3] for Matrix!(float, 3, 1)
alias Vector3(U) = Vector!(U, 3); // 2. resolves [U] > [float] with the result from 1

Vector3!U cross(U)(Vector3!U a, Vector3!U b) // 3. resolves [U] > [float] with the result from 2
{
    return Vector3!U.init;
}

void main()
{
    Vector3!float vv = cross(Vector3!float(), Vector3!float());
}

```

# The limitation

The alias template has no overloads and one single `TemplateInstance` must be found in `AliasDeclaration`. More complicated cases will fail. E.g.

```D
alias SomeAlias1(T) = SomeTemp!T; // ok
alias SomeAlias2(T) = somemod.SomeTemp!T; // ok
/////////////////
// IFTI will fail for the rest
alias SomeAlias3(T) = SomeTemp!T.SomeInnerType; 
alias SomeAlias4(T, U) = somemod.SomeTemp!T.SomeInnerTemp!U;
alias SomeAlias5(T) = int;
```
The analogous code of the failed cases can't compile in C++ either. The correct resolution may not be as much desired and will be much more difficult to achieve.

Since the matching part is <del>simplified from</del> the same as the old, it does no more than whatever the old does. e.g.

```D
struct TempT(U) {}

void foo(U : U*)(TempT!U v) {} // nomatch

alias AliasT(U: U*) = TempT!U; // nomatch

void fee(U)(AliasT!U v);

void main()
{
 foo(TempT!(float)()); // error
 fee(AliasT!(float*)()); // error
}
```

----

Previous discussions and rationale:\
<https://github.com/dlang/dmd/pull/9778>\
<https://github.com/dlang/DIPs/blob/master/DIPs/other/DIP1023.md>

A fuller test can be found here:\
<https://github.com/Lucipetus/dmd_test/blob/master/atifti.d>\
With cases of aliases defined across modules, and of `isInstanceOf` from `std`.

### CODE REVIEW

**Comment by thewilsonator** [comment]
Please open a PR to `dlang/dlang.org` that temporarily disables those tests in the spec.

**Review by thewilsonator — COMMENTED** [review]
Thanks for tackling this issue. 
This is a heavily indented method and this adds a _lot_ more indentation. I've opened a [PR](https://github.com/dlang/dmd/pull/15087/files?diff=split&w=1) to reduce the indentation a bit, if you could rebase after that PR goes in, that would make reviewing this a lot easier and more able to see where (if at all) the indentation in the PR can be reduced.

**Comment by dkorpel** [comment]
> Please open a PR to `dlang/dlang.org` that temporarily disables those tests in the spec.

https://github.com/dlang/dlang.org/pull/3571 is merged now

**Comment by Lucipetus** [comment]
> > Please open a PR to `dlang/dlang.org` that temporarily disables those tests in the spec.
> 
> [dlang/dlang.org#3571](https://github.com/dlang/dlang.org/pull/3571) is merged now

Great! 

"ci/circleci" still fails though. I have no idea what that is.

**Comment by thewilsonator** [comment]
Hmm, that looks like code coverage failed to upload, network error maybe? restarted it.

**Comment by Lucipetus** [comment]
> Thanks for tackling this issue. This is a heavily indented method and this adds a _lot_ more indentation. I've opened a [PR](https://github.com/dlang/dmd/pull/15087/files?diff=split&w=1) to reduce the indentation a bit, if you could rebase after that PR goes in, that would make reviewing this a lot easier and more able to see where (if at all) the indentation in the PR can be reduced.

Sure. I still need to figure out how to reuse the old matching code, if it's possible, but I figure it's better to have other review the idea first.

File: compiler/src/dmd/dtemplate.d [hunk 1]
```diff
@@ -4130,6 +4129,162 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
...
+                                    case TY.Tinstance:
+                                        atdi = cast(TypeInstance) atd.getType();
+                                        asc = s._scope;
+                                        break;
+                                    default:
+                                        break;
+                                    }
+                                    if (atdi && asc)
+                                    {
+                                        Objects* adedtypes = new Objects(td.parameters.length);
```
Comment (maxhaton) [inline compiler/src/dmd/dtemplate.d:4165]: `adedtypes` does not spark joy (variable name)

File: compiler/src/dmd/dtemplate.d [hunk 2]
```diff
@@ -4130,6 +4129,162 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
...
+                                        adedtypes.zero();
+                                        // alias SomeAlias(<td.parameters>) = SomeTempOrAlias!(<adedtypes>>);
+                                        MATCH am = deduceType(t, asc, atdi, td.parameters, adedtypes);
+                                        if (am != MATCH.exact)
+                                            goto Lnomatch;
+                                        // flatten adedtypes
```
Comment (maxhaton) [inline compiler/src/dmd/dtemplate.d:4171]: `flatten` (infinitive) suggests it may be possible to separate this out into at least a nested function.

File: compiler/test/compilable/testatifti.d [hunk 3]
```diff
@@ -0,0 +1,108 @@
+module testatifti;
```
Comment (maxhaton) [inline compiler/test/compilable/testatifti.d:1]: A note to an issue number and an explanation of what ATIFTI is here, please.

File: compiler/src/dmd/dtemplate.d (hunk 1 above)
Comment (Lucipetus) [reply compiler/src/dmd/dtemplate.d:4165]: I prefixed the usual names seen in the context with 'a' (for alias). I can't think of a better name yet, any suggestions?

File: compiler/test/compilable/testatifti.d (hunk 3 above)
Comment (Lucipetus) [reply compiler/test/compilable/testatifti.d:3]: Done.

**Comment by Lucipetus** [comment]
> Thanks for tackling this issue. This is a heavily indented method and this adds a _lot_ more indentation. I've opened a [PR](https://github.com/dlang/dmd/pull/15087/files?diff=split&w=1) to reduce the indentation a bit, if you could rebase after that PR goes in, that would make reviewing this a lot easier and more able to see where (if at all) the indentation in the PR can be reduced.

I've made `L2` into a function `resolveTemplateInstantiation`. There will be conflicts~

**Comment by Lucipetus** [comment]
I just set the PR state to `ready`. 
I made `L2` (where the deduction happens) into a function `resolveTemplateInstantiation`. The old functionality should be unaffected (if I didn't blunder), whether I'm using it right for the inserted alias template resolution is to be reviewed.

**Comment by Lucipetus** [comment]
I think the limitation of my solution is well defined: it works on simple alias template that expands to a template class/struct instance or another alias template. I've put examples and descriptions in the changelog.

This decade long issue remains unsolved: [1653](https://issues.dlang.org/show_bug.cgi?id=1653) - basically `alias cstr = const(T)[]`.

Again, my solution is really, really simple, it won't take more than one minute to see: all I do is call the existing matching code to resolve the parameters from the bottom up recursively. It shouldn't break any existing code, unless somehow, someone relies on deduction not working at all for alias templates. 

If I should first author a DIP to gain more feedback, please let me know. I did not do that because I thought this topic had already went through discussions and I wanted to get my code reviewed first to see if it was too naive and not worth the time - now that it's been two weeks with little progress, I think it's either this or because a DIP is required.

**Comment by PetarKirov** [comment]
@WalterBright what do you think?

**Comment by rikkimax** [comment]
Hey @RazvanN7, do you think you can annoy @WalterBright to please have a read? :P

File: compiler/src/dmd/dtemplate.d [hunk 4]
```diff
@@ -4129,6 +4128,65 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
                                         goto L2;
                                 }
                             }
+                            if (TemplateDeclaration td = s.isTemplateDeclaration())
```
Comment (WalterBright) [inline compiler/src/dmd/dtemplate.d:4131]: This is very hard to review. I suggest taking the code from 4132 .. 4189 and pulling it out into a separate function, which includes documentation for what it does.

File: compiler/src/dmd/dtemplate.d [hunk 5]
```diff
@@ -4141,185 +4199,192 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
...
+            }
+            visit(cast(Type)t);
+            return;
+
+        Lnomatch:
+            //printf("no match\n");
+            result = MATCH.nomatch;
+        }
+
+        private bool resolveTemplateInstantiation(Objects* tiargs, Objects* tdtypes, TemplateDeclaration tempdecl, TypeInstance tp, Objects* dedtypes)
```
Comment (WalterBright) [inline compiler/src/dmd/dtemplate.d:4213]: This function appears to be code cut&pasted out of the original. But it is very difficult to verify it. I suggest that the large refactoring go in a separate PR as a predecessor to this PR.

**Review by WalterBright — CHANGES_REQUESTED** [review]
It's really hard to see what is a semantic change and what is refactoring. The refactoring should be done in a separate, preceding PR.

File: compiler/src/dmd/dtemplate.d (hunk 5 above)
Comment (Lucipetus) [reply compiler/src/dmd/dtemplate.d:4213]: You are right. I've opened a new PR: https://github.com/dlang/dmd/pull/15263

File: compiler/src/dmd/dtemplate.d [hunk 6]
```diff
@@ -4129,6 +4128,37 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
                                         goto L2;
                                 }
                             }
+                            if (TemplateDeclaration td = s.isTemplateDeclaration())
+                            {
```
Comment (PetarKirov) [inline compiler/src/dmd/dtemplate.d:4132]: Can we merge the consecutive `if` statements since they have the same condition?
```suggestion

```

Comment (Lucipetus) [reply compiler/src/dmd/dtemplate.d:4132]: 4121 iters through the overloads of `td` and reassign `td` with them. Merging the two would require some modification between 4121 ~ 4130. I think we can revisit this later.

Comment (PetarKirov) [reply compiler/src/dmd/dtemplate.d:4132]: You're right, not sure how I overlooked this.

File: compiler/src/dmd/dtemplate.d (hunk 4 above)
Comment (Lucipetus) [reply compiler/src/dmd/dtemplate.d:4131]: > This is very hard to review. I suggest taking the code from 4132 .. 4189 and pulling it out into a separate function, which includes documentation for what it does.

I've pulled out part of it. Would you have another look?
